File: backend/api/routes/oculomotor.py
"""
backend/api/routes/oculomotor.py

Oculomotor routes for CogniSense unified backend.
"""

from fastapi import APIRouter, HTTPException, Query

from services.oculomotor_analyzer import get_oculomotor_analyzer
from utils.oculomotor_models import (
    OculomotorAnalysisRequest,
    OculomotorAnalysisResponse,
    OculomotorTaskStartResponse,
)

router = APIRouter()


@router.get("/oculomotor/health")
def oculomotor_health() -> dict:
    analyzer = get_oculomotor_analyzer()
    return analyzer.health_status()


@router.post("/oculomotor/analyze", response_model=OculomotorAnalysisResponse)
def analyze_oculomotor(request: OculomotorAnalysisRequest) -> OculomotorAnalysisResponse:
    analyzer = get_oculomotor_analyzer()
    try:
        return analyzer.analyze_metrics(request.metrics, source=request.source)
    except Exception as exc:
        raise HTTPException(status_code=500, detail=f"Oculomotor analysis failed: {exc}") from exc


@router.post("/oculomotor/analyze/demo", response_model=OculomotorAnalysisResponse)
def analyze_oculomotor_demo() -> OculomotorAnalysisResponse:
    analyzer = get_oculomotor_analyzer()
    return analyzer.generate_demo_result()


@router.get("/oculomotor/report/latest", response_model=OculomotorAnalysisResponse)
def analyze_oculomotor_latest_report() -> OculomotorAnalysisResponse:
    analyzer = get_oculomotor_analyzer()
    try:
        return analyzer.analyze_latest_report()
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except Exception as exc:
        raise HTTPException(status_code=500, detail=f"Failed to load latest oculomotor report: {exc}") from exc


@router.post("/oculomotor/task/start", response_model=OculomotorTaskStartResponse)
def start_oculomotor_camera_task(
    timeout_seconds: int = Query(default=420, ge=120, le=1800),
) -> OculomotorTaskStartResponse:
    analyzer = get_oculomotor_analyzer()
    try:
        payload = analyzer.run_standalone_task(timeout_seconds=timeout_seconds)
        return OculomotorTaskStartResponse(**payload)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except RuntimeError as exc:
        raise HTTPException(status_code=500, detail=str(exc)) from exc
    except Exception as exc:
        raise HTTPException(status_code=500, detail=f"Failed to launch camera task: {exc}") from exc

File: src/services/oculomotor.js
/**
 * src/services/oculomotor.js
 *
 * Oculomotor analysis API client for CogniSense unified backend.
 */

const ML_API_URL = (import.meta.env.VITE_ML_API_URL || 'http://localhost:8000').trim()

async function parseError(response) {
  let detail = 'Unknown error'
  try {
    const payload = await response.json()
    detail = payload?.detail || JSON.stringify(payload)
  } catch {
    detail = await response.text()
  }
  return `API ${response.status}: ${detail}`
}

export async function analyzeOculomotorMetrics(metrics, source = 'manual') {
  const response = await fetch(`${ML_API_URL}/api/oculomotor/analyze`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ metrics, source }),
  })

  if (!response.ok) {
    throw new Error(await parseError(response))
  }

  return response.json()
}

export async function analyzeOculomotorDemo() {
  const response = await fetch(`${ML_API_URL}/api/oculomotor/analyze/demo`, {
    method: 'POST',
  })

  if (!response.ok) {
    throw new Error(await parseError(response))
  }

  return response.json()
}

export async function analyzeLatestOculomotorReport() {
  const response = await fetch(`${ML_API_URL}/api/oculomotor/report/latest`, {
    method: 'GET',
  })

  if (!response.ok) {
    throw new Error(await parseError(response))
  }

  return response.json()
}

export async function startOculomotorCameraTask(timeoutSeconds = 420) {
  const timeout = Number.isFinite(Number(timeoutSeconds)) ? Number(timeoutSeconds) : 420
  const response = await fetch(`${ML_API_URL}/api/oculomotor/task/start?timeout_seconds=${Math.round(timeout)}`, {
    method: 'POST',
  })

  if (!response.ok) {
    throw new Error(await parseError(response))
  }

  return response.json()
}

export async function getOculomotorHealth() {
  const response = await fetch(`${ML_API_URL}/api/oculomotor/health`, {
    method: 'GET',
  })

  if (!response.ok) {
    throw new Error(await parseError(response))
  }

  return response.json()
}
